### ISSUE TITLE
Unable to understand "If you use the row index as a unique identifier, you need to make sure that new data gets appended to the end of the dataset, and no row ever gets deleted" on page 50

### ISSUE THREAD

**Issue by pranavpa8788** [ageron/handson-ml3#62]
I actually had doubt regarding "But both these solutions will break next time you fetch an updated dataset" which is referring to the methods of creating test set using numpy's random.permutation method and creating, loading the test set. 

Now, I understand that even if we use a global seed when using numpy's permutation method in order to the same numbers, it'll cause issues if our data gets rearranged as it will cause the ML algorithm to see more data than its supposed to.

And creating the test set and loading is kind of a static approach as we will have to manually create it again and again everytime there is a change in the data set.

The solution suggested is to use the hash method. Now the part I'm struggling to understand is, if the data changes order here, we still have the same issue right (assuming the index column remains unchanged) or else these index values would need to be attached/fixed for each row instance. I guess we can counter that by using some sort of id for each instance

Now coming to my actual question, "If you use the row index as a unique identifier, you need to make sure that new data gets appended to the end of the dataset, and no row ever gets deleted", I understand here that if we insert new data in between, it'll jumble up the old data which can end in the wrong place (test data getting into training data and vice-versa). But why is it suggested to avoid deleting data? 

If we are properly loading the csv file and the hash function isn't even dependent on our data's size, why then would this be an issue? If it's because the data (row) has already been fed to our model, isn't that more to do with our model than an issue with using row index as unique identifier?

I saw a similar question here: https://stackoverflow.com/questions/62229411/unique-identifier-in-a-dataset-index-problems
But I don't understand the answer given there

**Comment by pranavpa8788** [comment]
> If you delete a row it will mess up the index column (unless you do a reset_index()) and when you run a loop it won't find that index number and throw an error.

But if I consider this entire process to be as a "create train test set" function, wouldn't I call that kind of function when my data set changes (deletion of row) anyways?

**Comment by Giovanniko** [comment]
Yes I believe you would, but I think the author is just pointing this out to help those that may not think about that. The audience of the book is quite broad, which make it a great resource.

**Comment by pranavpa8788** [comment]
I see, that makes sense. I suppose due to the issue of only being able to append rows at the end is a deal breaker for using row based hashing anyways. Thanks @Giovanniko for clarifying and thank you @ageron for this book! I'm loving it so far
